### PR TITLE
otp: echo actor and kvcounter demo actor

### PR DESCRIPTION
**Pull request by stevelr**
Signed-off-by: steve <dev@somecool.net>

### CODE REVIEW

File: actor/echo/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+serde_json ="1.0"
+serde = {version = "1.0", features = ["derive"]}
+
+[dependencies.wasmcloud-interface-httpserver]
+version = "0.1"
+path = "../../../interfaces/httpserver/rust"
+
+[dependencies.wasmbus-rpc]
+version = "0.3"
+path = "../../../weld/rpc-rs"
```
Comment (brooksmtownsend) [inline actor/echo/Cargo.toml:25]: If I read this repo correctly, it seems to depend on the interfaces repository. I think this may be better as a git dependency at least until we have published versions of these crates

File: actor/echo/Makefile [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# examples/actor/echo
+
+KEYDIR ?= .keys
+CARGO ?= cargo
+REVISION := 8
```
Comment (brooksmtownsend) [inline actor/echo/Makefile:5]: We can restart the REVISION to 0 or 1 for the example

File: actor/echo/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+A simple http echo server
+This example demonstrates the actor code generation and
+a handler for the http-server interface.
+
+To build, `make` and `make run`
+
+To see the echo, curl a local url such as `curl localhost:8080/abc`
```
Comment (brooksmtownsend) [inline actor/echo/README.md:7]: Is this still accurate information? I don't see a `make run` rule, which may be a more complicated setup w/ the OTP runtime

File: actor/echo/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+use wasmcloud_interface_httpserver::{HttpRequest, HttpResponse, HttpServer, HttpServerReceiver};
+
+#[derive(Debug, Default, Actor, HealthResponder)]
+#[services(Actor, HttpServer)]
+struct EchoActor {}
+
+/// Implementation of HttpServer trait methods
+#[async_trait]
+impl HttpServer for EchoActor {
+    async fn handle_request(&self, _ctx: &Context, value: &HttpRequest) -> RpcResult<HttpResponse> {
```
Comment (brooksmtownsend) [inline actor/echo/src/lib.rs:12]: This looks awesome 🎉

File: actor/kvcounter/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+version = "0.2"
+path = "../../../interfaces/keyvalue/rust"
+
+[dependencies.wasmcloud-interface-httpserver]
+version = "0.1"
+path = "../../../interfaces/httpserver/rust"
+
+[dependencies.wasmbus-rpc]
+version = "0.3"
+path = "../../../weld/rpc-rs"
```
Comment (brooksmtownsend) [inline actor/kvcounter/Cargo.toml:31]: same comment as above re: weld/interfaces

File: actor/kvcounter/README.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# K/V Counter
+
+This actor accepts `GET` requests and for each URL it gets, 
+it will increment a call counter for that URL and return the result
```
Comment (brooksmtownsend) [inline actor/kvcounter/README.md:4]: I think this may read better as 
```
... it will increment counter based on that URL path and ...
```

Comment (brooksmtownsend) [reply actor/kvcounter/README.md:4]: It's a nitpick for sure, but I think it makes it a bit clearer what the actor is doing

File: actor/kvcounter/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+/// increment the counter by the amount, returning the new value
+async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
+    let tx = WasmHost::to_provider("wasmcloud:keyvalue", "default")?;
+    let kv = KeyValueSender::new(&tx);
+
+    let new_val = kv.increment(ctx, &IncrementRequest { key, value }).await?;
+    Ok(new_val as i32)
+}
```
Comment (brooksmtownsend) [inline actor/kvcounter/src/lib.rs:46]: I've been looking at this block of code for a little while as the steps feel like they could be consolidated a bit, and cleaned up for the user. There's nothing that needs to be done now, but I have a few questions / suggestions
1. Could we skip the `tx` step where we create a `WasmHost` and simply give the `KeyValueSender::new` a contract  ID and link definition?
2. Along the same lines, could we pass the context into the `KeyValueSender::new` constructor or is context supposed to be operation specific (e.g. is there a use case for multiple contexts for multiple KeyValueSenders`

If both of the above optimizations could be done, that would change this code to be:
```
/// increment the counter by the amount, returning the new value
async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
    let kv = KeyValueSender::new(ctx, "wasmcloud:keyvalue", "default")?;
    let new_val = kv.increment(&IncrementRequest { key, value }).await?;
    Ok(new_val as i32)
}
``` 

It's not much of a difference here, but from what I see it looks like these steps could certainly be consolidated. I also dislike always having `ctx` as the first parameter but I'll admit this is a common pattern and I haven't used it much.

For this PR, I don't feel like this has to be done so this is mostly comment/question material, want to make sure I'm not making assumptions

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
A couple small suggestions and a comment/question

File: actor/echo/README.md (hunk 3 above)
Comment (stevelr) [reply actor/echo/README.md:7]: you're right. I'll remove the comment. We need to decide on how to document running actors, whether through the web ui or with a future manifest file.

File: actor/echo/Cargo.toml (hunk 1 above)
Comment (stevelr) [reply actor/echo/Cargo.toml:25]: submitted issues for getting those repos set up with CI for publishing to crates.io

Comment (stevelr) [reply actor/echo/Cargo.toml:25]: I have published some of the common dependencies, to avoid having to use relative paths. I should have commented these out here.

File: actor/kvcounter/src/lib.rs (hunk 7 above)
Comment (stevelr) [reply actor/kvcounter/src/lib.rs:46]: Those are good questions which I will address in a different forum, since they are longish-answers and I'm sure others will ask about it later. Here's the TL;DR:

The short answer to the first one is that we could consolidate those two lines and save a line of code. Since it's only one line of code for the actor developer, I don't think it rises to a requirement for MVP.

ctx is in part a hook to incorporate environmental or other data that you might need, and it doesn't create a breaking api change if you add those later.

Putting Context into the constructor would save the ctx parameter on ServiceXSender api calls, but we still need it on the ServiceXReceiver interfaces, because providers need it to know the actor that sent the call. Right now, both *Sender and *Receiver implement the same ServiceX trait. It would be possible to generate different traits for sender and receiver, but even then, actors that act as receiver for a trait would still see a ctx parameter on every handler. I think it might still be a good idea to experiment with, to see what the resulting apis look like. This would be a big change to codegen, and would cause some destabilization if we did it before 0.50.

Comment (brooksmtownsend) [reply actor/kvcounter/src/lib.rs:46]: Thank you for the TL;DR, agreed that the little optimizations aren't a requirement for MVP. Definitely want to avoid premature optimizations that may require us to do something like adding in the `ctx` parameter later on. I think we'll know better from the feedback we get from people who play with the new interfaces / actor API what we need to improve on

File: actor/echo/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+authors = [ "wasmcloud Team" ]
+edition = "2018"
+
+[lib]
+crate-type = ["cdylib", "rlib"]
+
+[dependencies]
+async-trait = "0.1"
+console_log = "0.2"
+futures = "0.3"
```
Comment (autodidaddict) [inline actor/echo/Cargo.toml:13]: I know there's nothing we can do about this now, but it still feels like fingernails on a chalkboard to see a single-threaded webassembly module require the use of the `futures` crate. :(

File: actor/echo/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+log = "0.4"
+serde_bytes = "0.11"
+serde_json ="1.0"
+serde = {version = "1.0", features = ["derive"]}
+
+[dependencies.wasmcloud-interface-httpserver]
+version = "0.1"
+#path = "../../../interfaces/httpserver/rust"
```
Comment (autodidaddict) [inline actor/echo/Cargo.toml:21]: Please remove comments like this as they're likely to confuse first-time users.

File: actor/echo/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+[dependencies.wasmbus-rpc]
+version = "0.3"
+#path = "../../../weld/rpc-rs"
```
Comment (autodidaddict) [inline actor/echo/Cargo.toml:25]: Remove potentially confusing comments.

File: actor/echo/README.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+A simple http echo server
```
Comment (autodidaddict) [inline actor/echo/README.md:1]: By github convention, READMEs generally have a heading, so please just put a `# ` markdown header indicator in front of this line.

File: actor/echo/Makefile [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+wasm: build/$(PROJECT)_s.wasm
+
+$(DIST_WASM) : $(TARGET)/release/$(PROJECT)_s.wasm
+	@mkdir -p $(dir $@)
+	cp -p $< $@
+
+
+.ONESHELL:
+build/stub.exs: $(DIST_WASM)
```
Comment (autodidaddict) [inline actor/echo/Makefile:29]: please please tell me that we're not requiring the use of elixir script execution as part of the core actor development experience. I can't tell what this is for from the surrounding context, so if we don't need this it should be removed.

File: actor/kvcounter/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+futures = "0.3"
+log = "0.4"
+env_logger = "0.9"
+serde_bytes = "0.11"
+serde_json ="1.0"
+serde = {version = "1.0", features = ["derive"]}
+
+[dependencies.wasmcloud-interface-keyvalue]
+version = "0.2"
+#path = "../../../interfaces/keyvalue/rust"
```
Comment (autodidaddict) [inline actor/kvcounter/Cargo.toml:23]: Remove comments

File: actor/kvcounter/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+[dependencies.wasmcloud-interface-httpserver]
+version = "0.1"
+#path = "../../../interfaces/httpserver/rust"
+
+[dependencies.wasmbus-rpc]
+version = "0.3"
+#path = "../../../weld/rpc-rs"
```
Comment (autodidaddict) [inline actor/kvcounter/Cargo.toml:31]: Remove comments

File: actor/kvcounter/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            status_code,
+            ..Default::default()
+        };
+        Ok(resp)
+    }
+}
+
+/// increment the counter by the amount, returning the new value
+async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
+    let tx = WasmHost::to_provider("wasmcloud:keyvalue", "default")?;
```
Comment (autodidaddict) [inline actor/kvcounter/src/lib.rs:41]: The 0.18 and earlier versions of the actor SDKs all made the default case the easiest. If you're only ever using the `default` link name, then you should simply be able to call `KeyValueSender::new()` and it should assume the use of `default`. We should only have to call `WasmHost::to_provider("...", "...") if we want to change the link name. 

Further, requiring a developer to manually type the contract ID shouldn't be required, this should be made available to them as a constant in the crate they rely on for the key-value interface. `KeyValueSender` comes from the previously generated library, so it should already know its contract ID and not put the burden on the developer for typing that.

File: actor/kvcounter/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+console_log = "0.2"
+form_urlencoded = "1.0"
+futures = "0.3"
+log = "0.4"
+env_logger = "0.9"
+serde_bytes = "0.11"
+serde_json ="1.0"
+serde = {version = "1.0", features = ["derive"]}
+
+[dependencies.wasmcloud-interface-keyvalue]
```
Comment (autodidaddict) [inline actor/kvcounter/Cargo.toml:21]: Convert this to the far more readable dependency statement:
```
wasmcloud-interface-keyvalue = "0.2"
```

**Review by autodidaddict — CHANGES_REQUESTED** [review]
My biggest issue is that we're forcing developers to enter in the contract ID when the generated code absolutely knows that contract. I don't particularly have a preference for _how_ we accomplish it, but an actor developer should be able to create `KeyValueHost::new()` and have it default to the `default` link name and the `wasmcloud:keyvalue` contract ID. 

I understand that we might be forced to deal with creating both a `KeyValueHost` and a `WasmCloud::to_provider` , but that feels very much like we've forced developers to bear the burden of our own internal code architecture. 

In short, _it was easier to write an actor in 0.18_ and we need to fix that. At the very least, the `CONTRACT_ID` should be exposed as a public const in each of these modules if we require the use of the `to_provider` method. In an ideal world, developers don't have to make a second instance of something just to bear the `to_provider` output.

File: actor/echo/Makefile (hunk 12 above)
Comment (stevelr) [reply actor/echo/Makefile:29]: This was a band-aid I was using while waiting for manifests to become available for OTP.

File: actor/kvcounter/src/lib.rs (hunk 15 above)
Comment (stevelr) [reply actor/kvcounter/src/lib.rs:41]: fixed

**Comment by stevelr** [comment]
I don't seem to have permissions to merge this. Feedback from brooks and kevin has been incorporated.

**Comment by autodidaddict** [comment]
One of us needed to re-review. Every time you commit new changes to an approved PR it dismisses the approval

File: actor/kvcounter/src/lib.rs (hunk 15 above)
Comment (autodidaddict) [reply actor/kvcounter/src/lib.rs:41]: This doesn't appear to be fixed... you're still using the magic string "wasmcloud:keyvalue" and supplying "default" explicitly when, if you don't need anything other than the default, you shouldn't have to supply one (worse case scenario make people supply `None`, best case, the `to_provider` function doesn't take a link name, and you have a `to_provider_with_linkname` that takes the link name.

Comment (autodidaddict) [reply actor/kvcounter/src/lib.rs:41]: What if we have the following signatures:
* `WasmHost::to_provider(keyvalue::CONTRACT_ID)` (link name is "default")
* `WasmHost::to_provider_with_linkname(keyvalue::CONTRACT_ID, "florbleghast")` (explicit link name)

**Review by autodidaddict — CHANGES_REQUESTED** [review]

File: build/makefiles/actor.mk [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+# -----------
+# PROJECT      - Short name for the project, must be valid filename chars, no spaces
+# CLAIMS       - Space-separtaed list of capability contracts to use for signing
+#                These should match the capability providers the actor needs to use.
+#                For example: 
+#                    wasmcloud:httpserver wasmcloud:builtin:logging
+# VERSION      - The actor version number, usually semver format, X.Y.Z
+# REVISION     - A number that should be incremented with every build,
+#                whether or not VERSION has changed
+# REG_URL      - Registry url, e.g. 'localhost;5000' or 'wasmcloud.azurecr.io'
```
Comment (brooksmtownsend) [inline build/makefiles/actor.mk:15]: This semicolon `localhost;5000` should be a colon, `localhost:5000`

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
Tiny request, and I agree with @autodidaddict's review that we should make the default case as simple as possible. Not supplying `"default"` is a plus (previously we had `keyvalue::new('other_link')` and `keyvalue::default()`, but we should definitely avoid supplying the contract ID manually as that's prone to errors. I could see someone confusing `wasmcloud:http_server` with `wasmcloud:httpserver` and getting frustrated.

I still would like to see us move towards my suggestion after MVP (consolidating the WasmHost and KeyValueSender constructors into a single line, e.g. the actor developer avoiding instantiating the host) but contract ID as a const is a must

File: build/makefiles/actor.mk (hunk 17 above)
Comment (stevelr) [reply build/makefiles/actor.mk]: fixed

**Comment by stevelr** [comment]
This change was in other repositories but I'll mention here since it was required to fix the kvcounter actor.
All services that were declared in the smithy file with a capability contract id now have a contract_id() "static method" on the service trait to get the capability contract id. For actors sending to providers, you can use a `new()` constructor that builds the transport with the right capability contract and "default" link. There's also a `new_with_link("..")` constructor that lets you override the default link name. There's a third constructor that lets you override both contract and link name, but actors aren't likely to use that.

File: actor/kvcounter/src/lib.rs [hunk 18]
```diff
@@ -38,9 +38,8 @@ impl HttpServer for KvCounterActor {
 
 /// increment the counter by the amount, returning the new value
 async fn increment_counter(ctx: &Context, key: String, value: i32) -> RpcResult<i32> {
-    let tx = WasmHost::to_provider("wasmcloud:keyvalue", "default")?;
-    let kv = KeyValueSender::new(&tx);
-
-    let new_val = kv.increment(ctx, &IncrementRequest { key, value }).await?;
-    Ok(new_val as i32)
+    let new_val = KeyValueSender::new()
```
Comment (autodidaddict) [inline actor/kvcounter/src/lib.rs:41]: :tada: This is what I'm talking about. Very clean.

**Review by autodidaddict — APPROVED** [review]

**Review by brooksmtownsend — APPROVED** [review]
